File: nim/src/dsa/data_structures/hash_table.nim
## Hash Table - Open addressing with linear probing
##
## Time Complexity (average):
## - put: O(1)
## - get: O(1)
## - remove: O(1)
## - contains: O(1)

import std/hashes

type
  EntryState = enum
    Empty, Occupied, Deleted

  Entry[K, V] = object
    key: K
    value: V
    state: EntryState

  HashTable*[K, V] = object
    buckets: seq[Entry[K, V]]
    count: int
    capacity: int

const
  InitialCapacity = 16
  LoadFactorThreshold = 0.75

proc newHashTable*[K, V](): HashTable[K, V] =
  ## Create a new empty hash table
  result.capacity = InitialCapacity
  result.buckets = newSeq[Entry[K, V]](InitialCapacity)
  result.count = 0

proc findSlot[K, V](ht: HashTable[K, V], key: K): int =
  ## Find the slot for a key
  let h = hash(key) and (ht.capacity - 1)
  var index = h
  var firstDeleted = -1
  while true:
    case ht.buckets[index].state
    of Empty:
      if firstDeleted != -1:
        return firstDeleted
      return index
    of Deleted:
      if firstDeleted == -1:
        firstDeleted = index
    of Occupied:
      if ht.buckets[index].key == key:
        return index
    index = (index + 1) and (ht.capacity - 1)
    if index == h:
      break
  if firstDeleted != -1:
    return firstDeleted
  return -1

proc resize[K, V](ht: var HashTable[K, V]) =
  ## Resize the hash table
  let oldBuckets = ht.buckets
  ht.capacity = ht.capacity * 2
  ht.buckets = newSeq[Entry[K, V]](ht.capacity)
  ht.count = 0
  for entry in oldBuckets:
    if entry.state == Occupied:
      let index = ht.findSlot(entry.key)
      ht.buckets[index] = Entry[K, V](key: entry.key, value: entry.value, state: Occupied)
      inc ht.count

proc put*[K, V](ht: var HashTable[K, V], key: K, value: V) =
  ## Insert or update a key-value pair
  if float(ht.count + 1) / float(ht.capacity) > LoadFactorThreshold:
    ht.resize()
  let index = ht.findSlot(key)
  if ht.buckets[index].state != Occupied:
    inc ht.count
  ht.buckets[index] = Entry[K, V](key: key, value: value, state: Occupied)

proc `[]=`*[K, V](ht: var HashTable[K, V], key: K, value: V) =
  ## Insert or update using []= operator
  ht.put(key, value)

proc get*[K, V](ht: HashTable[K, V], key: K): V =
  ## Get value by key
  ## Raises KeyError if key not found
  let index = ht.findSlot(key)
  if index == -1 or ht.buckets[index].state != Occupied or ht.buckets[index].key != key:
    raise newException(KeyError, "Key not found")
  ht.buckets[index].value

proc getOrDefault*[K, V](ht: HashTable[K, V], key: K, default: V): V =
  ## Get value by key, or return default if not found
  let index = ht.findSlot(key)
  if index == -1 or ht.buckets[index].state != Occupied or ht.buckets[index].key != key:
    return default
  ht.buckets[index].value

proc `[]`*[K, V](ht: HashTable[K, V], key: K): V =
  ## Get value using [] operator
  ht.get(key)

proc contains*[K, V](ht: HashTable[K, V], key: K): bool =
  ## Check if key exists
  let index = ht.findSlot(key)
  index != -1 and ht.buckets[index].state == Occupied and ht.buckets[index].key == key

proc hasKey*[K, V](ht: HashTable[K, V], key: K): bool =
  ## Check if key exists (alias for contains)
  ht.contains(key)

proc remove*[K, V](ht: var HashTable[K, V], key: K): bool =
  ## Remove a key-value pair, returns true if found
  let index = ht.findSlot(key)
  if index == -1 or ht.buckets[index].state != Occupied or ht.buckets[index].key != key:
    return false
  ht.buckets[index].state = Deleted
  dec ht.count
  true

proc delete*[K, V](ht: var HashTable[K, V], key: K) =
  ## Remove a key-value pair
  ## Raises KeyError if key not found
  if not ht.remove(key):
    raise newException(KeyError, "Key not found")

proc isEmpty*[K, V](ht: HashTable[K, V]): bool =
  ## Check if the hash table is empty
  ht.count == 0

proc size*[K, V](ht: HashTable[K, V]): int =
  ## Return the number of entries
  ht.count

proc len*[K, V](ht: HashTable[K, V]): int =
  ## Return the number of entries (alias for size)
  ht.count

proc clear*[K, V](ht: var HashTable[K, V]) =
  ## Remove all entries
  ht.buckets = newSeq[Entry[K, V]](ht.capacity)
  ht.count = 0

proc keys*[K, V](ht: HashTable[K, V]): seq[K] =
  ## Get all keys
  result = @[]
  for entry in ht.buckets:
    if entry.state == Occupied:
      result.add(entry.key)

proc values*[K, V](ht: HashTable[K, V]): seq[V] =
  ## Get all values
  result = @[]
  for entry in ht.buckets:
    if entry.state == Occupied:
      result.add(entry.value)

iterator pairs*[K, V](ht: HashTable[K, V]): (K, V) =
  ## Iterate over key-value pairs
  for entry in ht.buckets:
    if entry.state == Occupied:
      yield (entry.key, entry.value)

when isMainModule:
  var ht = newHashTable[string, int]()
  ht["one"] = 1
  ht["two"] = 2
  ht["three"] = 3
  assert ht["one"] == 1
  assert ht["two"] == 2
  assert ht.contains("three")
  assert not ht.contains("four")
  assert ht.size() == 3
  assert ht.remove("two")
  assert not ht.contains("two")
  assert ht.size() == 2
  echo "HashTable tests passed!"

File: python/src/dsa/data_structures/hash_table.py
"""Hash table implementation using open addressing with linear probing.

Time: O(1) average for insert, get, remove
Space: O(n)
"""

from typing import Generic, TypeVar, Optional, Iterator
from dataclasses import dataclass
from enum import Enum

K = TypeVar("K")
V = TypeVar("V")

INITIAL_CAPACITY = 16
LOAD_FACTOR = 0.75


class _EntryState(Enum):
    EMPTY = 0
    DELETED = 1
    OCCUPIED = 2


@dataclass
class _Entry(Generic[K, V]):
    key: Optional[K]
    value: Optional[V]
    state: _EntryState


class HashTable(Generic[K, V]):
    """A generic hash table implementation with open addressing."""

    def __init__(self, capacity: int = INITIAL_CAPACITY) -> None:
        self._capacity = max(capacity, INITIAL_CAPACITY)
        self._entries: list[_Entry[K, V]] = [
            _Entry(None, None, _EntryState.EMPTY) for _ in range(self._capacity)
        ]
        self._len = 0

    def _hash(self, key: K) -> int:
        return hash(key) % self._capacity

    def _resize(self) -> None:
        old_entries = self._entries
        self._capacity *= 2
        self._entries = [
            _Entry(None, None, _EntryState.EMPTY) for _ in range(self._capacity)
        ]
        self._len = 0

        for entry in old_entries:
            if entry.state == _EntryState.OCCUPIED:
                self.insert(entry.key, entry.value)  # type: ignore

    def insert(self, key: K, value: V) -> Optional[V]:
        """Insert a key-value pair. Returns the old value if key existed."""
        if (self._len + 1) / self._capacity > LOAD_FACTOR:
            self._resize()

        index = self._hash(key)
        start_index = index

        while True:
            entry = self._entries[index]

            if entry.state in (_EntryState.EMPTY, _EntryState.DELETED):
                self._entries[index] = _Entry(key, value, _EntryState.OCCUPIED)
                self._len += 1
                return None

            if entry.state == _EntryState.OCCUPIED and entry.key == key:
                old_value = entry.value
                self._entries[index] = _Entry(key, value, _EntryState.OCCUPIED)
                return old_value

            index = (index + 1) % self._capacity
            if index == start_index:
                raise RuntimeError("Hash table is full")

    def get(self, key: K) -> Optional[V]:
        """Get the value for a key, or None if not found."""
        index = self._hash(key)
        start_index = index

        while True:
            entry = self._entries[index]

            if entry.state == _EntryState.EMPTY:
                return None

            if entry.state == _EntryState.OCCUPIED and entry.key == key:
                return entry.value

            index = (index + 1) % self._capacity
            if index == start_index:
                return None

    def remove(self, key: K) -> Optional[V]:
        """Remove a key and return its value, or None if not found."""
        index = self._hash(key)
        start_index = index

        while True:
            entry = self._entries[index]

            if entry.state == _EntryState.EMPTY:
                return None

            if entry.state == _EntryState.OCCUPIED and entry.key == key:
                value = entry.value
                self._entries[index] = _Entry(None, None, _EntryState.DELETED)
                self._len -= 1
                return value

            index = (index + 1) % self._capacity
            if index == start_index:
                return None

    def contains(self, key: K) -> bool:
        """Return True if the key exists."""
        return self.get(key) is not None

    def is_empty(self) -> bool:
        """Return True if the table is empty."""
        return self._len == 0

    def __len__(self) -> int:
        """Return the number of elements."""
        return self._len

    def __getitem__(self, key: K) -> Optional[V]:
        """Get item by key."""
        return self.get(key)

    def __setitem__(self, key: K, value: V) -> None:
        """Set item by key."""
        self.insert(key, value)

    def __contains__(self, key: K) -> bool:
        """Check if key exists."""
        return self.contains(key)

    def keys(self) -> Iterator[K]:
        """Iterate over keys."""
        for entry in self._entries:
            if entry.state == _EntryState.OCCUPIED:
                yield entry.key  # type: ignore

    def values(self) -> Iterator[V]:
        """Iterate over values."""
        for entry in self._entries:
            if entry.state == _EntryState.OCCUPIED:
                yield entry.value  # type: ignore

    def items(self) -> Iterator[tuple[K, V]]:
        """Iterate over key-value pairs."""
        for entry in self._entries:
            if entry.state == _EntryState.OCCUPIED:
                yield entry.key, entry.value  # type: ignore

    def clear(self) -> None:
        """Remove all elements."""
        self._entries = [
            _Entry(None, None, _EntryState.EMPTY) for _ in range(self._capacity)
        ]
        self._len = 0
